### ISSUE TITLE
Support both AccurateRip V1 and AccurateRip V2 at the same time

### ISSUE THREAD

**Issue by JoeLametta** [JoeLametta/whipper#18]
This one is kindly being worked on by [MerlijnWajer](https://github.com/MerlijnWajer).

https://github.com/thomasvs/morituri/issues/119#issuecomment-172170456

**Comment by peaveyman** [comment]
Has there been progress made on this?

**Comment by JoeLametta** [comment]
@MerlijnWajer told me he already completed this one but I'm still waiting for him to reply my last e-mails.

**Comment by peaveyman** [comment]
Thanks!!!

**Comment by MerlijnWajer** [comment]
I'm going to see if I can prepare some initial patches over the next week.

**Comment by JoeLametta** [comment]
@MerlijnWajer Thanks for the reply.
Didn't want to put pressure on you.

**Comment by MerlijnWajer** [comment]
Clearly didn't manage to do it that soon. Really hoping to send some initial patches your way soon.

**Comment by JoeLametta** [comment]
@MerlijnWajer Great!

**Comment by MerlijnWajer** [comment]
I have submitted an initial PR for the tool used to calculate both V1 and V2 checksums. Once that works fine, the next step is to integrate V2 in the report/logger code, the offset code and the image code.

**Comment by MerlijnWajer** [comment]
Since the code to calculate V1 and V2 will be in Very Soon™, this is a next good thing to have. I have written some code to do this, but it's outside of morituri's code base, so I need to figure out how to integrate it. OTOH, this should be fairly straightforward:

1) Add the right fields to the result structure
2) For each track, calculcate both checksums
3) Perform some clever logic to decide if a track matches. (Some decisions need to be made there, I would suggest that if ARv1 *or* ARv2 matches, it's good enough)
4) Add code to the logger to print this extra info (I have code that probably applies 1:1 here)

**Comment by JoeLametta** [comment]
> I would suggest that if ARv1 or ARv2 matches, it's good enough)

+1 for the OR

PS: If you think I can help you completing this one just let me know.

**Comment by RecursiveForest** [comment]
This has since landed with https://github.com/JoeLametta/whipper/pull/37 , no?

**Comment by MerlijnWajer** [comment]
Correct. It only uses v1 still. But the task can support both v1 and v2. So, as my post above, it shouldn't a lot of work to add using ARv1 and ARv2 (half a day or so, with testing)

**Comment by JoeLametta** [comment]
@MerlijnWajer
Is there any status update available for this one?

**Comment by MerlijnWajer** [comment]
Yep. I'm working on it right now. Will see if I can get something ready for merging on friday.

**Comment by JoeLametta** [comment]
> Yep. I'm working on it right now. Will see if I can get something ready for merging on friday.

OK, thanks.

**Comment by ArchangeGabriel** [comment]
Just as a thought, I think it would be nice to have both ARv1 and v2 in the log file. Or at least to make it configurable. ;)

**Comment by MerlijnWajer** [comment]
@ArchangeGabriel - that is exactly what will happen. This code has been mostly ready for some time, but I am first getting the flac/encode code in. By the way, this should trivial for others to pick up too. So if someone wants to pick it up, I can just drop my current code here and guide someone to actually finish it. Otherwise - wait for the flac+tagging code to be done.

**Comment by MerlijnWajer** [comment]
I've done most of the work, but need to test it still (it's getting close to 3am)

This commit addr retreiving/storing/matching the ARv1 *AND* ARv2 checksums: https://github.com/MerlijnWajer/whipper/commit/feaf3ff7a268f23fce27b1bf5a15af28da4b4451

It doesn't yet modify result/logger.py to 'print the right thing'. And that will have to be done for every logger, too. So that's some work that I do not intend to do all by myself.

tl;dr commit I linked might work for both v1 and v2 matching, but is not complete, needs testing (by me), but once that is done, I need help to update all our loggers. I can get this part done tomorrow.

@JoeLametta could you help with that (loggers)?

**Comment by gorgobacka** [comment]
@MerlijnWajer @JoeLametta Are there any updates for this?

**Comment by MerlijnWajer** [comment]
Somebody (forgot the nickname) said they would make the loggers ARv1/ARv2 capable, so I'm kind of waiting for someone to do that. I can also pick this up later for one of the loggers.

**Comment by Scott-Bowman** [comment]
I hacked together ARv2 support; it's in my fork on the branch [SupportARv2](https://github.com/Scott-Bowman/whipper/tree/SupportARv2). I mainly did this for fun and my approach to implementing it probably isn't ideal but I'm kind of proud so I thought I'd  share. I've ripped about a dozen CD's with it and have not encountered any issues so far.

**Comment by RecursiveForest** [comment]
https://github.com/JoeLametta/whipper/pull/187

**Comment by gorgobacka** [comment]
I will test it on some CDs.

**Comment by JoeLametta** [comment]
This issue can now be closed because #187 (RecursiveForest) has been merged.
